Checkum policy blocks prefix list

cstamas opened the pull request:
Fixes
http://issues.sonatype.org/browse/NXCM-5188

Multiple fixes:
- NPE fixed that is apparent in log in the issue (is general, not prefix file related!)
- Implemented a fix: override per-request the ChecksumPolicy in Maven repositories
- making use of the core change in `RemotePrefixFileStrategy`, added UT
- added comments to remove temptation (like I had)

CI
https://bamboo.zion.sonatype.com/browse/NXO-OSSF1-1

cstamas committed:
NXCM-5188: NPE fix

In the issue attached log, it's clearly visible that validation
did reject the item (policy=Strict but no hashes were present).
Later, an NPE is thrown, as the cleanup method is not protected
from this case.

In all other case it would work as contentValid would be
TRUE for all of those. This is the only branch in logic
where it is FALSE, and the remoteHash is null too.

In other words, cleanup is needed when content is
invalid, but hashes arrived into cache.

cstamas committed:
NXCM-5188: Added checksum relax possibility and using it in AR strategy

Added simple "per request" ChecksumPolicy override feature to maven
specific repositories (as ChecksumPolicy itself is maven specific).
Using it one can "relax" the policy for single request.

Making use of it in RemotePrefixFileStrategy, to relax
the policy if its STRICT to one step below (STRICT_IF_EXISTS).

Added UT that validates the change. UT will fail if relevant
changes are commented out from RemotePrefixFileStrategy,

cstamas committed:
NXCM-5188: Removed temptation by adding inline explanation

I was tempted to change these methods, but finally figured
it's completely unneeded.

Comment explains why.

cstamas committed:
NXCM-5188: Typo in comment

cstamas commented:
CI +1'd
https://bamboo.zion.sonatype.com/browse/NXO-OSSF1-1

nabcos commented:
+1

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/routing/internal/RemotePrefixFileStrategy.java:
```diff
@@ -147,8 +148,14 @@ protected StorageFileItem retrieveFromRemoteIfExists( final MavenProxyRepository
-        request.getRequestContext().put( Manager.ROUTING_INITIATED_FILE_OPERATION_FLAG_KEY, Boolean.TRUE );
         request.setRequestRemoteOnly( true );
+        request.getRequestContext().put( Manager.ROUTING_INITIATED_FILE_OPERATION_FLAG_KEY, Boolean.TRUE );
+        if ( ChecksumPolicy.STRICT == mavenProxyRepository.getChecksumPolicy() )
```
Why not unconditionally use ChecksumPolicy.IGNORE? Not a big deal, but extra remote request for SHA1/MD5 files that almost certainly will not be there and do not add much even if they are, seems pointless.

cstamas replied:
They will be there if remote is Nexus 2.4+, and they are certainly not there if remote is "static repo" like Central is, unless... we modify the already changed specs and make `prefixes.txt.sha1` a requirement for example. For transport corruption the checksum is still the best and simplest, and if it's already available, why not use it?

ifedorenko replied:
because this complicates the code and repo-to-repo interactions ever so slightly, and I highly doubt this improves reliability of prefix file processing in any meaningful ways (all transport errors will most likely be caught by standard tcp/ip logic). anyways, I don't feel strong enough about this, merge it as is if you think checksum validation is useful here.

/me shrugs

ifedorenko commented:
+1 if you agree that ChecksumPolicy.IGNORE is the right checksum policy to use for prefix files.

cstamas committed:
NXCM-5188: Checksum policy set to IGNORE for prefix requests

As the transport integrity here has small chances to slip in
as prefix file will be processed, and hence, checked/validated
